Source/WebCore/rendering/RenderLayer.cpp

```diff
@@ -3768,13 +3768,17 @@ void RenderLayer::paintLayerContents(GraphicsContext& context, const LayerPainti
     };
 
     if (paintingInfo.paintBehavior.contains(PaintBehavior::DraggableSnapshot) && paintingInfo.subtreePaintRoot) {
-        if (paintingInfo.subtreePaintRoot->hasLayer()) {
-            CheckedPtr subtreeRootLayer = paintingInfo.subtreePaintRoot->enclosingLayer();
-            bool isLayerInSubtree = (this == subtreeRootLayer) || isDescendantOf(*subtreeRootLayer);
+        auto isNotSelfReference = paintingInfo.subtreePaintRoot != &renderer();
+        auto hasLayer = paintingInfo.subtreePaintRoot->hasLayer();
 
-            if (isLayerInSubtree && (paintingInfo.subtreePaintRoot != &renderer() && shouldExcludeBasedOnContainingBlock()))
-                shouldPaintContent = false;
-        } else if (renderer().isAbsolutelyPositioned() && paintingInfo.subtreePaintRoot != &renderer() && shouldExcludeBasedOnContainingBlock()) {
+        CheckedPtr subtreeRootLayer = hasLayer ? paintingInfo.subtreePaintRoot->enclosingLayer() : nullptr;
+        bool isLayerInSubtree = subtreeRootLayer && ((this == subtreeRootLayer) || isDescendantOf(*subtreeRootLayer));
+
+        if (hasLayer && !isLayerInSubtree && isNotSelfReference) {
```

Comment:
This logic seems overly complicated, and the variables being tested are not independent; if this layer's renderer is the subtree root, we know that the renderer has a layer (|this|).

We need to deal with two issues here. First, `subtreePaintRoot->enclosingLayer()` can return the layer of some ancestor renderer, which might contain positioned descendants that are not descendants of subtreePaintRoot.

Second, `subtreePaintRoot` may have descendants whose containing blocks are ancestors of `subtreePaintRoot`, and should thus not show in the snapshot. Let's write the code to be clear about those two cases; use lambda with names if necessary.